### PR TITLE
Refactor `SentTransactionHash` to be a stricter type

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

The `SentTransactionHash` type previously used a `String` field to represent the transaction's hash in hexadecimal encoding. However, this has the risk that the type could be constructed with an invalid string (non-hexadecimal or not representative of a hash).

## Solution

Use the `hex::serialize` method to serialize a `transaction::Hash` field. This ensures that the hexadecimal string is only generated for serialization and an invalid string is never serialized.

The downside is that `transaction::Hash` had to have some new implementations for the traits `ToHex` and `FromHex`. I also refactored the `Display` and `Debug` implementations to avoid having repeated code that could risk becoming different in different traits.

## Review

@teor2345 and @dconnolly originally suggested the usage of `transaction::Hash` in `SentTransactionHash`.

### Reviewer Checklist

  - [x] Code implements Specs and Designs
  - [x] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

**Review by teor2345 — APPROVED** [review]
I like this change, let's do something similar for the other hex-encoding RPCs.

File: zebra-rpc/src/methods.rs [hunk 1]
```diff
@@ -188,4 +191,4 @@ pub struct GetBlockChainInfo {
 /// Response to a `sendrawtransaction` RPC request.
 ///
 /// A JSON string with the transaction hash in hexadecimal.
-pub struct SentTransactionHash(String);
+pub struct SentTransactionHash(#[serde(with = "hex")] transaction::Hash);
```
Comment (jvff) [inline zebra-rpc/src/methods.rs:192]: I managed to shorten the attribute and also make it future-proof if we end up having to `derive(Deserialize)` for it.

File: zebra-chain/src/transaction/hash.rs [hunk 2]
```diff
@@ -100,20 +101,55 @@ impl From<&Hash> for [u8; 32] {
...
+    /// Return the hash bytes in an order suitable for printing out byte by byte.
+    fn bytes_in_display_order(&self) -> [u8; 32] {
         let mut reversed_bytes = self.0;
         reversed_bytes.reverse();
-        f.write_str(&hex::encode(&reversed_bytes))
+        reversed_bytes
+    }
+}
+
+impl ToHex for &'_ Hash {
```
Comment (dconnolly) [inline zebra-chain/src/transaction/hash.rs]: Do we need this lifetime when it's a `&Hash`?

File: zebra-chain/src/transaction/hash.rs [hunk 3]
```diff
@@ -100,20 +101,55 @@ impl From<&Hash> for [u8; 32] {
     }
 }
 
-impl fmt::Display for Hash {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+impl Hash {
+    /// Return the hash bytes in an order suitable for printing out byte by byte.
```
Comment (dconnolly) [inline zebra-chain/src/transaction/hash.rs]:
```suggestion
    /// Return the hash bytes in big-endian byte-order suitable for printing out byte by byte.
    ///
    /// Zebra displays transaction and block hashes in big-endian byte-order,
    /// following the u256 convention set by Bitcoin and zcashd.
```

**Comment by dconnolly** [comment]
@jvff looks great! i made some small doc suggestions so haven't reviewed yet to trigger the merge but ready to do so

**Review by dconnolly — DISMISSED** [review]

**Review by teor2345 — DISMISSED** [review]
Looks good to me, I'm also interested in the question about the lifetime.

File: zebra-chain/src/transaction/hash.rs (hunk 3 above)
Comment (jvff) [reply zebra-chain/src/transaction/hash.rs]: Fixed, thanks!

**Review by dconnolly — DISMISSED** [review]

**Comment by teor2345** [comment]
@jvff would you mind resolving conflicts on this PR?

**Review by teor2345 — APPROVED** [review]

**Pull request merged**